apps/dashboard/src/components/html-editor.tsx

```diff
@@ -1,37 +1,68 @@
-import { useCallback, useMemo, useRef } from 'react';
-import { Completion, CompletionContext } from '@codemirror/autocomplete';
-import { EditorView } from '@uiw/react-codemirror';
+import { MutableRefObject, useCallback, useMemo, useRef } from 'react';
+import { Completion, CompletionContext, CompletionSource } from '@codemirror/autocomplete';
+import { EditorView, Extension } from '@uiw/react-codemirror';
 import { liquid, liquidCompletionSource } from '@codemirror/lang-liquid';
 import { html, htmlCompletionSource } from '@codemirror/lang-html';
 import { tags as t } from '@lezer/highlight';
 import { format } from 'prettier/standalone';
 import * as parserHtml from 'prettier/plugins/html';
 import * as parserLiquid from '@shopify/prettier-plugin-liquid/standalone';
 import { RiCodeSSlashFill } from 'react-icons/ri';
+import { JSONSchema7 } from 'json-schema';
 
-import { VariableEditor } from '@/components/primitives/variable-editor';
-import { useWorkflow } from '@/components/workflow-editor/workflow-provider';
-import { useParseVariables } from '@/hooks/use-parse-variables';
+import { CompletionRange, VariableEditor } from '@/components/primitives/variable-editor';
 import { cn } from '@/utils/ui';
 import { CompletionOption } from '@/utils/liquid-autocomplete';
 import { Tooltip } from '@/components/primitives/tooltip';
 import { TooltipContent } from '@/components/primitives/tooltip';
 import { TooltipTrigger } from '@/components/primitives/tooltip';
-import { useSaveForm } from '@/components/workflow-editor/steps/save-form-context';
 import { showErrorToast } from '@/components/primitives/sonner-helpers';
+import { LiquidVariable } from '@/utils/parseStepVariables';
 
 type HtmlEditorProps = {
+  viewRef: MutableRefObject<EditorView | null>;
+  lastCompletionRef: MutableRefObject<CompletionRange | null>;
   value: string;
+  variables: LiquidVariable[];
+  isAllowedVariable: (variable: LiquidVariable) => boolean;
   onChange: (value: string) => void;
+  saveForm?: () => void;
+  completionSources?: CompletionSource[];
+  extensions?: Extension[];
+  children?: React.ReactNode;
+  isPayloadSchemaEnabled?: boolean;
+  digestStepName?: string;
+  getSchemaPropertyByKey?: (key: string) => JSONSchema7 | undefined;
+  onCreateNewVariable?: (variableName: string) => Promise<void>;
+  onManageSchemaClick?: (variableName: string) => void;
+  skipContainerClick?: boolean;
 };
 
 const gutterElementClassName =
   '[&_.cm-gutterElement]:flex [&_.cm-gutterElement]:items-center [&_.cm-gutterElement]:justify-end [&_.cm-gutterElement]:text-text-soft [&_.cm-gutterElement]:font-code [&_.cm-gutterElement]:text-code-sm [&_.cm-gutterElement>span]:h-full';
 
-export function HtmlEditor({ value, onChange }: HtmlEditorProps) {
-  const { step, digestStepBeforeCurrent } = useWorkflow();
-  const { variables, isAllowedVariable } = useParseVariables(step?.variables, digestStepBeforeCurrent?.stepId);
-  const { saveForm } = useSaveForm();
```

Comment:
I moved all workflow related logic outside of this component. Similar idea is to applied to many other components.